runtime-sdk/src/modules/consensus_accounts/types.rs

```diff
@@ -29,13 +29,68 @@ pub struct Withdraw {
 pub struct Delegate {
     pub to: Address,
     pub amount: token::BaseUnits,
+    #[cbor(optional)]
+    pub receipt: u64,
 }
 
 /// Undelegate into runtime call.
 #[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
 pub struct Undelegate {
     pub from: Address,
     pub shares: u128,
+    #[cbor(optional)]
+    pub receipt: u64,
+}
+
+/// Kind of receipt.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[repr(u8)]
+pub enum ReceiptKind {
+    #[default]
+    Invalid = 0,
+    Delegate = 1,
+    UndelegateStart = 2,
+    UndelegateDone = 3,
+}
+
+impl ReceiptKind {
+    /// Whether the receipt kind is valid.
+    pub fn is_valid(&self) -> bool {
+        !matches!(self, Self::Invalid)
+    }
+}
+
+/// Take receipt internal runtime call.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct TakeReceipt {
+    pub kind: ReceiptKind,
+    pub id: u64,
+}
+
+/// A receipt.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct Receipt {
```

Comment:
I prefer splitting this into multiple structs, as it is hard to know which fields are/can be set by different kinds.

Reply:
Yeah it is just slightly annoying with the current deserializer.